README.md
# HSolo

HSolo is a library for estimating similarity, affine, and projective homographies from matches produced by image feature detectors such as SURF and SIFT. HSolo will work for any image feature detector that produces an estimate of scale and rotation at each detected point.

![image](example_data/demo.png)

HSolo is unique because it can estimate a full 9-DOF homography bootstrapping from only a single correspondence. That makes it faster than most other estimators and it can also find solutions when the ratio of true matches in the set is extremely low. 

HSolo is written in C++ with Python and Matlab wrappers and is meant to be used as an alternative to OpenCV's `findHomography` and Matlab's `EstimateGeometricTransform2D`. 

For more information, please refer to the [arXiv paper](https://arxiv.org/pdf/2009.05004.pdf). SAND2020-9405

## Installation

### Requirements

* General:  CMake 3.8+, Eigen 3.3+, OpenCV 4.4+, Pybind11 2.5+
* Python Wrapper:  Python 3.7+
* Matlab Wrapper:  Matlab 2018a+
* C++ Compliers -- must support C++17
    * For Windows, we successfuly built using [Microsoft Build Tools for VS 2019](https://visualstudio.microsoft.com/downloads/#build-tools-for-visual-studio-2019 ).  Make sure all commands are run from a `Native Tools Command Prompt`.

#### Pybind11 Dependency

Pybind11 2.5+ should be added directly to the project directory as follows:

* Clone this repo and navigate into its root directory.
* Create `lib` directory and navigate into it.
* Run `git clone https://github.com/pybind/pybind11.git`

### C++ Static Library and Example Executable

#### Unix and MacOS

* Create and navigate to `build` directory under `HSolo`.
* Run `cmake -DBUILD_STATIC_LIBRARY=ON -DBUILD_EXAMPLE_EXECUTABLE=ON ..`
* Run `make`

#### Windows

* Create and navigate to `build` directory under `HSolo`.
* Run `cmake -G "Visual Studio 16 2019" -A x64 -DBUILD_STATIC_LIBRARY=ON -DBUILD_EXAMPLE_EXECUTABLE=ON -DEIGEN3_INCLUDE_DIR="<path to eigen directory>" -DOpenCV_DIR="<path to opencv directory>\build" ..`
* Run `cmake --build <path to HSolo>\build --config Release --target ALL_BUILD`

### Python Wrapper

#### Unix and MacOS
* In the `HSolo` directory, run `pip install .`

#### Windows

* Create and navigate to `build` directory under `HSolo`
* Run `cmake -G "Visual Studio 16 2019" -A x64 -DBUILD_PYTHON_WRAPPER=ON -DEIGEN3_INCLUDE_DIR="<path to eigen directory>" -DOpenCV_DIR="<path to opencv directory>\build" ..`
* Run `cmake --build <path to HSolo>\build --config Release --target ALL_BUILD`
* Run `SET HSOLO_INSTALL_USE_PREBUILT=1`
* Run `pip install .`

### Matlab Wrapper

#### Unix and MacOS

* Create and navigate to `build` directory under `HSolo`.
* Run `cmake -DBUILD_MATLAB_MEX=ON ..`
* Run `make`

#### Windows

* Create and navigate to `build` directory under `HSolo`
* Run `cmake -G "Visual Studio 16 2019" -A x64 -BUILD_MATLAB_MEX=ON -DEIGEN3_INCLUDE_DIR="<path to eigen directory>" -DOpenCV_DIR="<path to opencv directory>\build" ..`
* Run `cmake --build <path to HSolo>\build --config Release --target ALL_BUILD`

## Testing HSolo

* Python: Use the `pyhsolo_example` Jupyter Notebook in the root directory.
* C++:  Run the `hsolo_example` executable created in the `bin` directory.  Note:  You need to be in the `bin` directory when running the example.
* Matlab:  Run the `hsolo_example.m` script in the `matlab` directory.

## Supported Transform Types

| Transform      | Description |  Function | Min Number Correspondences
| ----------- | ----------- | ----------- | :----:  |
| Similarity | Translation, Scale, Rotation   | findSimilarity | 2 |
| Affine | Translation, Scale, Rotation, Skew   | findAffine | 3 |
| Projective | Out of plane projection  | findHomography | 4 |
| Translation | Move  | findTranslation | 1 |

`HSolo` does not directly solve for Fundamental or Essential Matricies.  The `HSolo` Projective solver can be used as a pre-filter in inlier-sparse cases.  The inliers produced by `HSolo` can then be fed into existing solvers such as OpenCV `findFundamentalMatrix`. 

## API

### Parameters:

* **Confidence**:  (0, 1.0]. Default = 0.95.  Likelihood that a correct solution will be found.  This determines the number of iterations run based on a dynamically updating estimate of the inlier rate.  The higher the confidence, the longer runtime.  Note: to better match other Matlab geometry solvers, this is given as (0, 100] in the Matlab wrapper
* **Reprojection Error Threshold**:  > 0, Default = 2.0.  The maximum reprojection error (in pixels) for a correspondence to be considered an inlier.  This parameter is denoted `MaxDistance` in the Matlab wrapper.
* **Filtered Inlier Rate**:  (0, 1.0]. Default = 0.7.  The estimated inlier rate of the correspondence set after being filtered by the initial homography estimate.  This determines the number of inner RANSAC iterations.  Lower estimates will increase the likelihood of a correct result, but increase the run time.
* **Max Iterations**:  > 0.  Default=1000.  The maximum number of iterations allowed to run regardless of other parameter settings.
* **Run Inner RANSAC Thresh**:  > 0. Default = 25.  The highest median reprojection error of the filtered inlier set in order for the inner ransac estimation to run.  Higher thresholds will increase the likelihood of a correct result, but increase the run time.  The default of 25 works well generally, but this can be tuned for specific datasets.
* **Search Top N Filtered**:  (>= 4 homography, >=3 for affine, >=2 for similarity).  Default = 20.  The number of the top N filtered matches to be searched during the inner RANSAC step. 

### Input Data

* **x,y Pixel Location**
* **Features Scale**
* **Orientation in Degrees**


### C++ API: 
```c++
struct HSoloCorrespondence
{
    double pt1_x;     // x-coordinate for 1st correspondence
    double pt1_y;     // y-coordinate for 1st correspondence
    double pt2_x;     // x-coordinate for 2nd correspondence
    double pt2_y;     // y-coordinate for 2nd correspondence
    double pt1_ori;   // orientation for 1st correspondence
    double pt1_scale; // scale for 1st correspondence
    double pt2_ori;   // orientation for 2nd correspondence
    double pt2_scale; // scale for 2nd correspondence
};
```

``` c++
std::vector<HSoloCorrespondence> corrs;
// Populate correspondences 

// Define parameter configuration
HSoloConfiguration config;
config.setConfidence(conf);
config.setErrorThreshold(thresh);
config.setFilteredInlierRate(fir);
config.setMaxIterations(miters);
config.setRunInnerRANSACThresh(rirth);
config.setSearchTopNfiltered(searchTopN);

// Input: hsolo_matches is std::vector<HSoloCorrespondence>
// Output: HSoloResult
HSolo hsolo;
HSoloResult hsoloResult = hsolo.findHomography(hsolo_matches, config);
```
### Python API: 
``` python
# Input:  x1y1, x2y2 are [Nx2] numpy arrays of the locations of the point corresondences
#         scales1, scales2 are [Nx1] numpy arrays of the scales of the point corresondences
#         oris1, oris2 are [Nx1] numpy arrays of the orientation (in degrees) of the point corresondences
# Output: H is a tuple containing (homography matrix, inliers)
#         If get_stats=True then H contains (homography matrix, inliers, result_stats)
H = find_homography(x1y1, x2y2, scales1, scales2, oris1, oris2)
```
### Matlab API:
``` matlab
% Input: matchedPoints are features detected and matched using matlab. detectXXXXFeatures and matchFeatures
% Output: Projective2D object
estimateGeometricTransform2D(...
        matchedPoints1, ...
        matchedPoints2, ...
        transformType...
    );
```

## Contributing

Thank you for your interest in contributing to the success of HSolo. Please see our [Contributing Guide](CONTRIBUTING.md) to learn more. 

## Copyright

Copyright 2021 National Technology & Engineering Solutions of Sandia, LLC (NTESS). 
Under the terms of Contract DE-NA0003525 with NTESS, the U.S. Government retains certain rights in this software.

Licensed under the Apache License, Version 2.0


CHANGELOG.md
# Change Log

## [v1.0] - 2022-06-06

### Added

- Initial HSolo Open Source Release


src/python_bindings.cpp
// Copyright 2021 National Technology & Engineering Solutions of Sandia, LLC (NTESS). 
// Under the terms of Contract DE-NA0003525 with NTESS, the U.S. Government retains certain rights in this software.
//
//    Licensed under the Apache License, Version 2.0 (the "License");
//    you may not use this file except in compliance with the License.
//    You may obtain a copy of the License at

//        http://www.apache.org/licenses/LICENSE-2.0

//    Unless required by applicable law or agreed to in writing, software
//    distributed under the License is distributed on an "AS IS" BASIS,
//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//    See the License for the specific language governing permissions and
//    limitations under the License.

#include "HSolo.h"
#include <stdexcept>
#include <pybind11/pybind11.h>
#include <pybind11/stl_bind.h>
#include <pybind11/numpy.h>
#include <string>

namespace py = pybind11;

using namespace std;

std::vector<HSoloCorrespondence> prepareData(
    py::array_t<double> x1y1_,
    py::array_t<double> x2y2_,
    py::array_t<double> scales1_,
    py::array_t<double> scales2_,
    py::array_t<double> oris1_,
    py::array_t<double> oris2_,
    int min_sample_size)
{
    py::buffer_info x1y1_buffer = x1y1_.request();
    size_t x1y1_rows = x1y1_buffer.shape[0];
    size_t x1y1_cols = x1y1_buffer.shape[1];

    std::string minSampleSize = to_string(min_sample_size);

    if (x1y1_cols != 2)
    {
        throw std::invalid_argument("x1y1 should be an array with dims [n,2], n>=" + minSampleSize);
    }
    if (x1y1_rows < min_sample_size)
    {
        throw std::invalid_argument("x1y1 should be an array with dims [n,2], n>=" + minSampleSize);
    }

    py::buffer_info x2y2_buffer = x2y2_.request();
    size_t x2y2_rows = x2y2_buffer.shape[0];
    size_t x2y2_cols = x2y2_buffer.shape[1];

    if (x2y2_cols != 2)
    {
        throw std::invalid_argument("x2y2 should be an array with dims [n,2], n>=" + minSampleSize);
    }
    if (x2y2_rows != x1y1_rows)
    {
        throw std::invalid_argument("x1y1 and x2y2 should be the same size");
    }

    py::buffer_info scales1_buffer = scales1_.request();
    size_t scales1_rows = scales1_buffer.shape[0];
    size_t scales1_cols = scales1_buffer.shape[1];

    if (scales1_cols != 1)
    {
        throw std::invalid_argument("scales1 should be an array with dims [n,1], n>=" + minSampleSize);
    }
    if (scales1_rows != x1y1_rows)
    {
        throw std::invalid_argument("x1y1 and scales1 should be the same size");
    }

    py::buffer_info scales2_buffer = scales2_.request();
    size_t scales2_rows = scales2_buffer.shape[0];
    size_t scales2_cols = scales2_buffer.shape[1];

    if (scales2_cols != 1)
    {
        throw std::invalid_argument("scales2 should be an array with dims [n,1], n>=" + minSampleSize);
    }
    if (scales2_rows != x1y1_rows)
    {
        throw std::invalid_argument("x2y2 and scales2 should be the same size");
    }

    py::buffer_info oris1_buffer = oris1_.request();
    size_t oris1_rows = oris1_buffer.shape[0];
    size_t oris1_cols = oris1_buffer.shape[1];

    if (oris1_cols != 1)
    {
        throw std::invalid_argument("oris1 should be an array with dims [n,1], n>=" + minSampleSize);
    }
    if (oris1_rows != x1y1_rows)
    {
        throw std::invalid_argument("x1y1 and oris1 should be the same size");
    }

    py::buffer_info oris2_buffer = oris2_.request();
    size_t oris2_rows = oris2_buffer.shape[0];
    size_t oris2_cols = oris2_buffer.shape[1];

    if (oris2_cols != 1)
    {
        throw std::invalid_argument("oris2 should be an array with dims [n,1], n>=" + minSampleSize);
    }
    if (oris2_rows != x1y1_rows)
    {
        throw std::invalid_argument("x2y2 and oris2 should be the same size");
    }

    double *x1y1_ptr = (double *)x1y1_buffer.ptr;
    std::vector<double> x1y1;
    x1y1.assign(x1y1_ptr, x1y1_ptr + x1y1_buffer.size);

    double *x2y2_ptr = (double *)x2y2_buffer.ptr;
    std::vector<double> x2y2;
    x2y2.assign(x2y2_ptr, x2y2_ptr + x2y2_buffer.size);

    double *scales1_ptr = (double *)scales1_buffer.ptr;
    std::vector<double> scales1;
    scales1.assign(scales1_ptr, scales1_ptr + scales1_buffer.size);

    double *scales2_ptr = (double *)scales2_buffer.ptr;
    std::vector<double> scales2;
    scales2.assign(scales2_ptr, scales2_ptr + scales2_buffer.size);

    double *oris1_ptr = (double *)oris1_buffer.ptr;
    std::vector<double> oris1;
    oris1.assign(oris1_ptr, oris1_ptr + oris1_buffer.size);

    double *oris2_ptr = (double *)oris2_buffer.ptr;
    std::vector<double> oris2;
    oris2.assign(oris2_ptr, oris2_ptr + oris2_buffer.size);


    std::vector<HSoloCorrespondence> corrs;

    for (size_t ii = 0; ii < x1y1_rows; ii++)
    {
        HSoloCorrespondence corr;
        corr.pt1_x = x1y1[ii * 2];
        corr.pt1_y = x1y1[ii * 2 + 1];
        corr.pt2_x = x2y2[ii * 2];
        corr.pt2_y = x2y2[ii * 2 + 1];

        corr.pt1_scale = scales1[ii];
        corr.pt2_scale = scales2[ii];
        corr.pt1_ori = oris1[ii];
        corr.pt2_ori = oris2[ii];

        corrs.push_back(corr);
    }

    return corrs;
}

py::tuple prepareResult(HSoloResult& result, bool get_stats) {
    // no homography found, return None
    if (result.support == 0)
    {
        if (get_stats == true)
        {
            return py::make_tuple(py::none(), py::none(), py::none());
        }
        else
        {
            return py::make_tuple(py::none(), py::none());
        }
    }

    py::array_t<bool> inliers_ = py::array_t<bool>(result.inliers.size());
    py::buffer_info buf3 = inliers_.request();
    bool *ptr3 = (bool *)buf3.ptr;
    for (size_t i = 0; i < result.inliers.size(); i++)
        ptr3[i] = result.inliers[i];

    std::vector<float> H(9);
    H[0] = result.H(0, 0);
    H[1] = result.H(0, 1);
    H[2] = result.H(0, 2);
    H[3] = result.H(1, 0);
    H[4] = result.H(1, 1);
    H[5] = result.H(1, 2);
    H[6] = result.H(2, 0);
    H[7] = result.H(2, 1);
    H[8] = result.H(2, 2);

    py::array_t<float> H_ = py::array_t<float>({3, 3});
    py::buffer_info buf2 = H_.request();
    float *ptr2 = (float *)buf2.ptr;
    for (size_t i = 0; i < 9; i++)
        ptr2[i] = H[i];

    py::dict result_stats;
    result_stats["support"] = result.support;
    result_stats["score"] = result.score;
    result_stats["num_iters"] = result.num_iters;
    result_stats["num_inner_ransac_iters"] = result.num_inner_ransac_iters;

    if (get_stats == true)
    {
        return py::make_tuple(H_, inliers_, result_stats);
    }
    else
    {
        return py::make_tuple(H_, inliers_);
    }
}

py::tuple findTranslation_(py::array_t<double> x1y1_,
                      py::array_t<double> x2y2_,
                      py::array_t<double> scales1_,
                      py::array_t<double> scales2_,
                      py::array_t<double> oris1_,
                      py::array_t<double> oris2_,
                      double threshold,
                      double filtered_inlier_rate,
                      int search_top_n_filtered,
                      double conf,
                      int max_iters,
                      double run_inner_ransac_thresh,
                      double lo_det_limit,
                      double hi_det_limit,
                      bool get_stats,
                      bool refine_solution)
{
    try
    {
        HSoloConfiguration config;

        config.setErrorThreshold(threshold);
        config.setConfidence(conf);
        config.setFilteredInlierRate(filtered_inlier_rate);
        config.setRunInnerRANSACThresh(run_inner_ransac_thresh);
        config.setMaxIterations(max_iters);
        config.setSearchTopNfiltered(search_top_n_filtered);


        std::vector<HSoloCorrespondence> corrs = prepareData(x1y1_, x2y2_, 
                                        scales1_, scales2_,oris1_, oris2_, 3);

        HSolo hsolo;
        HSoloResult result = hsolo.findTranslation(corrs, config);

        return prepareResult(result, get_stats);
    }
    catch (const char *msg)
    {
        throw std::invalid_argument(msg);
    }
}


py::tuple findSimilairty_(py::array_t<double> x1y1_,
                          py::array_t<double> x2y2_,
                          py::array_t<double> scales1_,
                          py::array_t<double> scales2_,
                          py::array_t<double> oris1_,
                          py::array_t<double> oris2_,
                          double threshold,
                          double filtered_inlier_rate,
                          int search_top_n_filtered,
                          double conf,
                          int max_iters,
                          double run_inner_ransac_thresh,
                          bool get_stats,
                          bool refine_solution)
{
    try
    {

        HSoloConfiguration config;

        config.setErrorThreshold(threshold);
        config.setConfidence(conf);
        config.setFilteredInlierRate(filtered_inlier_rate);
        config.setRunInnerRANSACThresh(run_inner_ransac_thresh);
        config.setMaxIterations(max_iters);
        config.setSearchTopNfiltered(search_top_n_filtered);
        config.setRefineSolution(refine_solution); 

        std::vector<HSoloCorrespondence> corrs = prepareData(x1y1_, x2y2_, 
                                        scales1_, scales2_,oris1_, oris2_, 2);

        HSolo hsolo;
        HSoloResult result = hsolo.findSimilarity(corrs, config);

        return prepareResult(result, get_stats);
    }
    catch (const char *msg)
    {
        throw std::invalid_argument(msg);
    }
}

py::tuple findAffine_(py::array_t<double> x1y1_,
                      py::array_t<double> x2y2_,
                      py::array_t<double> scales1_,
                      py::array_t<double> scales2_,
                      py::array_t<double> oris1_,
                      py::array_t<double> oris2_,
                      double threshold,
                      double filtered_inlier_rate,
                      int search_top_n_filtered,
                      double conf,
                      int max_iters,
                      double run_inner_ransac_thresh,
                      bool get_stats,
                      bool refine_solution)
{
    try
    {
        HSoloConfiguration config;
        config.setErrorThreshold(threshold);
        config.setConfidence(conf);
        config.setFilteredInlierRate(filtered_inlier_rate);
        config.setRunInnerRANSACThresh(run_inner_ransac_thresh);
        config.setMaxIterations(max_iters);
        config.setSearchTopNfiltered(search_top_n_filtered);
        config.setRefineSolution(refine_solution);

        std::vector<HSoloCorrespondence> corrs = prepareData(x1y1_, x2y2_, 
                                        scales1_, scales2_,oris1_, oris2_, 3);

        HSolo hsolo;
        HSoloResult result = hsolo.findAffine(corrs, config);

        return prepareResult(result, get_stats);
    }
    catch (const char *msg)
    {
        throw std::invalid_argument(msg);
    }
}

py::tuple findHomography_(py::array_t<double> x1y1_,
                          py::array_t<double> x2y2_,
                          py::array_t<double> scales1_,
                          py::array_t<double> scales2_,
                          py::array_t<double> oris1_,
                          py::array_t<double> oris2_,
                          double threshold,
                          double filtered_inlier_rate,
                          int search_top_n_filtered,
                          double conf,
                          int max_iters,
                          double run_inner_ransac_thresh,
                          bool get_stats,
                          bool refine_solution)
{

    try
    {
        HSoloConfiguration config;
        config.setErrorThreshold(threshold);
        config.setConfidence(conf);
        config.setFilteredInlierRate(filtered_inlier_rate);
        config.setRunInnerRANSACThresh(run_inner_ransac_thresh);
        config.setMaxIterations(max_iters);
        config.setSearchTopNfiltered(search_top_n_filtered);
        config.setRefineSolution(refine_solution);

        std::vector<HSoloCorrespondence> corrs = prepareData(x1y1_, x2y2_, 
                                        scales1_, scales2_,oris1_, oris2_, 4);

        HSolo hsolo;
        HSoloResult result = hsolo.findHomography(corrs, config);

        return prepareResult(result, get_stats);
    }
    catch (const char *msg)
    {
        throw std::invalid_argument(msg);
    }
}


PYBIND11_MODULE(_pyhsolo, m)
{


    m.def("findHomography", &findHomography_, "Find a Homography from a set of n>=4 candidate correspondences.",
          py::arg("x1y1"),
          py::arg("x2y2"),
          py::arg("scales1"),
          py::arg("scales2"),
          py::arg("oris1"),
          py::arg("oris2"),
          py::arg("threshold") = 2.0,
          py::arg("filtered_inlier_rate") = 0.7,
          py::arg("search_top_n_filtered") = 20,
          py::arg("conf") = 0.95,
          py::arg("max_iters") = 1000,
          py::arg("run_inner_ransac_thresh") = 25.0,
          py::arg("get_stats") = true,
          py::arg("refine_solution") = true);

    m.def("findSimilarity", &findSimilairty_, "Find a Similarity Homography from a set of n>=2 candidate correspondences.",
          py::arg("x1y1"),
          py::arg("x2y2"),
          py::arg("scales1"),
          py::arg("scales2"),
          py::arg("oris1"),
          py::arg("oris2"),
          py::arg("threshold") = 2.0,
          py::arg("filtered_inlier_rate") = 0.7,
          py::arg("search_top_n_filtered") = 20,
          py::arg("conf") = 0.95,
          py::arg("max_iters") = 1000,
          py::arg("run_inner_ransac_thresh") = 25.0,
          py::arg("get_stats") = true,
          py::arg("refine_solution") = true);

    m.def("findAffine", &findAffine_, "Find a Affine Homography from a set of n>=3 candidate correspondences.",
          py::arg("x1y1"),
          py::arg("x2y2"),
          py::arg("scales1"),
          py::arg("scales2"),
          py::arg("oris1"),
          py::arg("oris2"),
          py::arg("threshold") = 2.0,
          py::arg("filtered_inlier_rate") = 0.7,
          py::arg("search_top_n_filtered") = 20,
          py::arg("conf") = 0.95,
          py::arg("max_iters") = 1000,
          py::arg("run_inner_ransac_thresh") = 25.0,
          py::arg("get_stats") = true,
          py::arg("refine_solution") = true);

    m.def("findTranslation", &findTranslation_, "Find a Translation from a set of n>=1 candidate correspondences.",
          py::arg("x1y1"),
          py::arg("x2y2"),
          py::arg("scales1"),
          py::arg("scales2"),
          py::arg("oris1"),
          py::arg("oris2"),
          py::arg("threshold") = 2.0,
          py::arg("filtered_inlier_rate") = 0.7,
          py::arg("search_top_n_filtered") = 20,
          py::arg("conf") = 0.95,
          py::arg("max_iters") = 1000,
          py::arg("run_inner_ransac_thresh") = 13.0,
          py::arg("low_det_limit") = 0.1,
          py::arg("high_det_limit") = 2000.0,
          py::arg("get_stats") = true,
          py::arg("refine_solution") = true);

}
